### DISCUSSION TITLE
How to properly handle scaling for export purposes

### DISCUSSION THREAD

**Discussion by apwebber** [pyqtgraph/pyqtgraph#2222] [Q&A]
I know pyqtgraph is aimed mainly at interactive applications, but in some cases it is easier to let the user export graphs rather than handling conversion to something like matplotlib. Currently with pyqtgraph, the export menu allows the user to export the image at a higher resolution than it is displayed on screen, but the result is not what the user might expect, which is an increase in 'dpi' instead of just stretching the canvas. Font sizes and line widths stay the same, the plot just gets bigger, resulting in unreadable text.

I have played around with a custom printer/painter, for example using QPainter.scale(2,2) will increase the dpi of a QChart. But with PyQtGraph, results were not very successful.

Is there a good way to do this?

**Comment by pijyoi** [comment]
I imagine it would require changing ImageExporter to add a dpi parameter and then calling QImage.setDevicePixelRatio().

RemoteGraphicsView and the Jupyter RemoteFrameBuffer code do this. Exporting is after all just offscreen rendering.

**Reply by apwebber** [reply]
Ah I had overlooked setDevicePixelRatio() ! It works, and of course you have to multiply width and height of the QImage by the same factor.

**Reply by apwebber** [reply]
One last piece of the puzzle is that all QPens must be set to non-cosmetic so that the width scales appropriately. From what I can tell there would be two approaches, one would be to make sure all QPens are created as non-cosmetic when plotting the graphs, and another would be to access all QGraphicScene.Items() during export, and ensure all QPens are non-cosmetic. Bit of a pain either way.

**Reply by apwebber** [reply]
Ok this is a bit more complicated than I first thought. Setting the pens to non-cosmetic when plotting (e.g. using mkPen()) makes them appear like ribbons because, I think, the widths are related to the plot co-ordinate system. If your x axis is goes from 40 to 60 and line width is 1, the line is very thick in the dimension of the x axis. But if the y axis goes from 1 to 2000, the line is very thin in the dimension of the y axis.

**Reply by apwebber** [reply]
Current implementation involves a function to manually scale QPen items according to the dpi/scale the user requests (only implemented for PlotDataItems so far):

```
def scale_pens(self, scale):
        
        for item in self.getScene().items():
            if isinstance(item, PlotDataItem):
                pen = item.opts['pen']

                '''Calling setPen creates the QPen object'''
                if not isinstance(pen, QtGui.QPen):
                    item.setPen(pen)
                    pen = item.opts['pen']

                width = pen.width()
                pen.setWidth(scale * width)
                item.setPen(pen)
```

The scene can then be rendered with the pens set to the correct width, without using `cosmetic = False`.

Afterwards, the scale_pens function needs to be called again with scale=1/scale to get back to the original scale, else the GUI will display the fat lines.

I'm doing all this in the exporter. I have a feeling this isn't the 'right' way to do it, and it should be done in the gui, so that the user can see what the result will be. Like a proper document/graphics app.

**Reply by apwebber** [reply]
The above works as intended the first time, but if the plot is re-drawn/updated then the lines revert to the scaled (fat) state. Not sure why since they were 'shrunk' back down. Even completely new plots with new data and after calling `GraphicsLayoutWidget.clear()` have fat lines. Even when lines are defined using `mkPen(width=1)`.
